Use separate BitSet cache in Doc Level Security

tvernum opened the pull request:
Document level security was depending on the shared
"BitsetFilterCache" which (by design) never expires its entries.

However, when using DLS queries - particularly templated ones - the
number (and memory usage) of generated bitsets can be significant.

This change introduces a new cache specifically for BitSets used in
DLS queries, that has memory usage constraints and access time expiry.

The whole cache is automatically cleared if the role cache is cleared.
Individual bitsets are cleared when the corresponding lucene index
reader is closed.

Closes: #30974

tvernum committed:
Use separate BitSet cache in Doc Level Security

Document level security was depending on the shared
"BitsetFilterCache" which (by design) never expires its entries.

However, when using DLS queries - particularly templated ones - the
number (and memory usage) of generated bitsets can be significant.

This change introduces a new cache specifically for BitSets used in
DLS queries, that has memory usage constraints and access time expiry.

The whole cache is automatically cleared if the role cache is cleared.
Individual bitsets are cleared when the corresponding lucene index
reader is closed.

elasticmachine commented:
Pinging @elastic/es-security

tvernum commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into dls-bitset-cache

tvernum committed:
Fix test

tvernum committed:
Merge branch 'master' into dls-bitset-cache

tvernum committed:
Change access time TTL to 1 week

tvernum commented:
I haven't written docs for these new cache config properties (`ttl` and `max_bytes`) but will do so on Monday.

I'm curious about other people's opinions on the default values for these.

- I've set `max_bytes` to 50Mb, because it seems like it would be a problem to use more than that on a small (1Gb) node, but I don't really have the information at hand to know whether that's big enough to store typical workloads or whether users will have cache thrashing.

- I set `ttl` to a week (7 * 24 hours) because I think we want to depend on the `max_bytes` setting to keep the cache under control, rather than aggresively expiring entries if they aren't used for a day or so.

I could easily be persuaded to a different set of values, but since the main purpose of this change it to reduce the amount of memory used by DLS bitsets, we don't want to make those numbers too big.

jpountz reviewed (changes requested):
We should try to find a better way to evict on closed segments. We used to have other caches that required linear scans for eviction and this caused issues, eg. when closing a node because then all its indices would close one after the other, and since close listeners are called synchonously from IndexReader#close, this would run in O(n^2) as a function of the number of segments on the node.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetBitsetCache.java:
```diff
@@ -0,0 +1,194 @@
+
+    @Override
+    public void onClose(IndexReader.CacheKey ownerCoreCacheKey) {
+        bitsetCacheHelper.removeKeysIf(key -> key.matchesIndex(ownerCoreCacheKey));
```
I'll look at maintaining a secondary mapping. The problem with the two-level cache is that we don't have good ways to manage expiry.
The lower level caches could have access-time expiry, but it's hard to do memory-size expiry.
The upper level caches can do memory-size expiry, but will expire everything under a  `IndexReader.CacheKey` rather than expiring the LRU queries.

tvernum committed:
Merge branch 'master' into dls-bitset-cache

# Conflicts:
#	x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/SecurityIndexSearcherWrapper.java
#	x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/SecurityIndexReaderWrapperIntegrationTests.java
#	x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/SecurityIndexSearcherWrapperUnitTests.java
#	x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

tvernum committed:
Treat null bitset marker as 0 bytes

tvernum committed:
Maintain map of index to cache-key

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetBitsetCache.java:
```diff
@@ -0,0 +1,207 @@
+
+/**
+ * This is a cache for {@link BitSet} instances that are used with the {@link DocumentSubsetReader}.
+ * It bounded by memory size and access time.
```
```suggestion
 * It is bounded by memory size and access time.
```

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetBitsetCache.java:
```diff
@@ -0,0 +1,207 @@
+     * cache is never filled.
+     */
+    static final Setting<TimeValue> CACHE_TTL_SETTING =
+        Setting.timeSetting("xpack.security.dls_fls.bitset.cache.ttl", TimeValue.timeValueHours(24 * 7), Property.NodeScope);
```
Should we create a new prefix `xpack.security.dls.` for these settings instead of reusing `xpack.security.dls_fls` ?

tvernum replied:
Maybe, would you prefer that?
It seems like a toss-up to me. Having too many different prefixes is potentially confusing for users, but it's also confusing to use a "_fls" prefix for something that's only for DLS.

I suspect most users don't really care as long as we document it, but I'm very happy to align this with your preference if you have one.

jkakavas replied:
I think I prefer being explicit and erring a little on the "too many prefixes" case.

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/support/CacheIteratorHelper.java:
```diff
@@ -0,0 +1,59 @@
+import java.util.function.Predicate;
+
+/**
+ * A utility class to facilitating iterating over (and modifying) a {@link org.elasticsearch.common.cache.Cache}.
```
```suggestion
 * A utility class to facilitate iterating over (and modifying) a {@link org.elasticsearch.common.cache.Cache}.
```

jkakavas reviewed (approved):
The changes / logic looks good to me, I just left some wording nits and a suggestion about the naming of the related settings. I don't feel particularly confident with this part but thankfully @jpountz makes up for it

jpountz reviewed (changes requested):
I left some minor comments but otherwise it looks good to me, including the mapping from queries to cache keys.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetBitsetCache.java:
```diff
@@ -0,0 +1,207 @@
+        Setting.timeSetting("xpack.security.dls_fls.bitset.cache.ttl", TimeValue.timeValueHours(24 * 7), Property.NodeScope);
+
+    static final Setting<ByteSizeValue> CACHE_BYTES_SETTING =
+        Setting.byteSizeSetting("xpack.security.dls_fls.bitset.cache.max_bytes",
```
Thanks, I'll update it.
I tried to follow the pattern from other security caches, but they are less consistent.
e.g.
- `xpack.security.authc.realms.<type>.<name>.cache.max_users`
- `xpack.security.authz.store.roles.cache.max_size`

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetBitsetCache.java:
```diff
@@ -0,0 +1,207 @@
+            // This ensures all insertions into the set are guarded by ConcurrentHashMap's atomicity guarantees.
+            keysByIndex.compute(indexKey, (ignore2, set) -> {
+                if (set == null) {
+                    set = new HashSet<>();
```
Sure, happy to change it.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/accesscontrol/DocumentSubsetBitsetCache.java:
```diff
@@ -0,0 +1,207 @@
+     * cache is never filled.
+     */
+    static final Setting<TimeValue> CACHE_TTL_SETTING =
+        Setting.timeSetting("xpack.security.dls_fls.bitset.cache.ttl", TimeValue.timeValueHours(24 * 7), Property.NodeScope);
```
If a role has a templated query, and the metadata changes (e.g. A user gets added to a new AD group) then there isn't anything that would trigger an eviction of those entries except closing lucene readers (segment merges), expiry time or the cache being full.
If we remove the default TTL then, for cold indices, we would rely exclusively on evicting when the cache is full. This might not happen (we might only use 95% of the memory limit) and we'd carry these bitsets forever (until the role cache is explicitly cleared, or the node is restarted).

That doesn't bother me too much - I don't think the scenario is particularly likely.  
But on a low-memory node, every byte counts so it feels like we should evict at some point.

Perhaps it's OK to never expire by default, and then have a recommendation that low memory environments explictly configure these settings.

jpountz replied:
So the case I was a bit worried about is the opposite one, where there are say 1000 role queries that are frequently reused. And every week there is suddenly a lot of load on the system because all bitsets need to be recreated at about the same time.

jpountz replied:
Hmm nevermind. What I said doesn't make sense, if the role queries were reused then they wouldn't be evicted. :)
+1 to keep this TTL

tvernum committed:
Switch to concurrent set

tvernum committed:
Merge branch 'master' into dls-bitset-cache

tvernum committed:
Address feedback

- Rename "max_bytes" to "size"
- Rename "dls_fls" to "dls"
- Fix typos in Javadoc

tvernum commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into dls-bitset-cache
